aio_general_context: flush: return short flush on error

bhalevy opened the pull request:
As seen in https://github.com/scylladb/seastar/issues/1784, having flush retry `io_submit` in a loop might be futile if the kernel internal queue is full and we need to reap completions to free up space for more iocb:s.

This change breaks out of the loop as soon as
`io_submit` fails and just moves the remaining entries in the iocbs queue into the front so more entries can be queued (otherwise, we'd need a circular queue that will penalize the error-free path for the unlikely event of an error)

Fixes scylladb/seastar#1784

bhalevy committed:
aio_general_context: flush: abort on unexpected errors

EAGAIN is expected here when "Insufficient resources are available to queue any iocbs" (see io_submit(2)).
Abort on any other error, as those indicate an internal error on our side.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

tchaikov commented on src/core/reactor_backend.cc:
```diff
@@ -337,26 +337,18 @@ void aio_general_context::queue(linux_abi::iocb* iocb) {
+        if (__builtin_expect(r <= 0, false)) {
+            // errno == EAGAIN is expected here, but we don't explicitly assert that.
+            // just adjust the queue and return the number of iocbs we were able to submit
+            std::memmove(iocbs.get(), begin, (last - begin) * sizeof(iocb*));
```
nit,
```c++
std::move(begin, last, iocbs.get());
```

StephanDollberg commented on src/core/reactor_backend.cc:
```diff
@@ -337,26 +337,18 @@ void aio_general_context::queue(linux_abi::iocb* iocb) {
-            retry_until = clock::now() + 1s;
-        } else {
-            assert(clock::now() < retry_until);
+        if (__builtin_expect(r <= 0, false)) {
```
Would still make sense to check/assert/retry for errors other than EAGAIN? 

From what I read from the man page most of those are probably logic errors.

tchaikov replied:
i'd vote for `assert()` . it's guarded by `__builtin_expect(r <= 0, false)`, so presumably this should not hurt the performance that much, considering that we always undefine `NDEBUG`.

bhalevy replied:
Can do.

bhalevy replied:
> r==0 should be impossible (only occurs if we try to submit 0 iocbs, which we don't don't). So perhaps we should include it in the assert case or have a separate assert case.

Done in [2cbb4cc](https://github.com/scylladb/seastar/commit/2cbb4cccec913f9a74c09e7bca88483175384cf4)

bhalevy commented:
In v2 ([2cbb4cc](https://github.com/scylladb/seastar/commit/2cbb4cccec913f9a74c09e7bca88483175384cf4)):
- use `std::move` rather than `memmove`
- verify `errno` and abort using `on_fatal_internal_error` if it is not `EAGAIN.`

bhalevy committed:
aio_general_context: flush: return short flush on error

As seen in https://github.com/scylladb/seastar/issues/1784, having flush retry `io_submit` in a loop might be futile if the kernel internal queue is full and we need to reap completions to free up space for more iocb:s.

This change breaks out of the loop as soon as `io_submit` fails and just moves the remaining entries in the iocbs queue into the front so more entries can be queued (otherwise, we'd need a circular queue that will penalize the error-free path for the unlikely event of an error).

Since a short flush can leave polling iocb:s queued in userspace, the aio backend must not enter a blocking wait while such pending iocb:s remain. Otherwise the reactor could sleep waiting for events even though the iocb that would wake it, such as the timerfd or smp wakeup eventfd poll, was never submitted to the kernel. Track that pending state and force the wait path to keep polling instead.

The same concern applies to the preemption context. If the hrtimer iocb remains queued in userspace, the aio ring cannot make need_preempt() true, so request_preemption() must not spin waiting for a completion that was not submitted.

Fixes scylladb/seastar#1784

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy commented:
Updated the branch after review feedback.

Changes since the previous push:
- folded the pending-IOCB sleep guard into the short-flush patch so the reactor does not enter a blocking wait while polling IOCBs remain queued in userspace
- folded the preemption-context guard into the same patch so request_preemption() does not spin waiting for an hrtimer IOCB that was not submitted
- kept the unexpected-error handling as the second patch
- updated the short-flush commit message to document both hazards

Local verification:
- git diff --check origin/master..HEAD
- ninja -C build/dev CMakeFiles/seastar.dir/src/core/reactor_backend.cc.o

bhalevy commented:
@xemul please review

StephanDollberg commented:
We have been running with the two original patches for a couple years now. Once you have those two in you'll quickly need this third one on top as the asserts just shift: https://github.com/redpanda-data/seastar/commit/f922a4bb5b0c06dc7bde48b14dfe2842efc03c81

Note there is a problem with all these patches that seems inherent to the way the reactor is built and how the aio/iocb stuff works.

These three patches handle the case well where there is a short temporary spike in IOCB demand. Instead of crashing we just churn through all the IOCBs.

However, what this patch series still doesn't handle is when there is constant IOCB demand higher than `max-networking-io-control-blocks` (many concurrent pending reads or wait_input_shutdown). In that case there is basically a constant oustanding queue which possibly only moves slowly. In all likely hood this results in a degraded system as important things like node heartbeats are queue behind other less important connections. Observability into this state is also very poor. This also combines poorly with `max-networking-io-control-blocks` silently getting truncated when max-aio-nr is low or smp count is high.

Nevertheless because the first scenario is not uncommon and it's much better to just absorb these spikes than crash we are still running with these patches. But it's not a fix-all state.

avikivity commented:
> We have been running with the two original patches for a couple years now. Once you have those two in you'll quickly need this third one on top as the asserts just shift: [redpanda-data@f922a4b](https://github.com/redpanda-data/seastar/commit/f922a4bb5b0c06dc7bde48b14dfe2842efc03c81)
> 
> Note there is a problem with all these patches that seems inherent to the way the reactor is built and how the aio/iocb stuff works.
> 
> These three patches handle the case well where there is a short temporary spike in IOCB demand. Instead of crashing we just churn through all the IOCBs.
> 
> However, what this patch series still doesn't handle is when there is constant IOCB demand higher than `max-networking-io-control-blocks` (many concurrent pending reads or wait_input_shutdown). In that case there is basically a constant oustanding queue which possibly only moves slowly. In all likely hood this results in a degraded system as important things like node heartbeats are queue behind other less important connections. Observability into this state is also very poor. This also combines poorly with `max-networking-io-control-blocks` silently getting truncated when max-aio-nr is low or smp count is high.

My recommendation is to try out the assymmetric io_uring backend [*]. It should not have any of these problems.

> 
> Nevertheless because the first scenario is not uncommon and it's much better to just absorb these spikes than crash we are still running with these patches. But it's not a fix-all state.


[*] the secret agenda is that you hit all the bugs first

bhalevy commented:
@avikivity what do you suggest doing with this change?

avikivity commented on src/core/reactor_backend.cc:
```diff
@@ -444,26 +444,19 @@ void aio_general_context::queue(linux_abi::iocb* iocb) {
 
 size_t aio_general_context::flush() {
```
> As seen in https://github.com/scylladb/seastar/issues/1784, having flush retry `io_submit` in a loop might be futile if the kernel internal queue is full and we need to reap completions to free up space for more iocb:s.

How is it possible? The number of IOCBs is equal to the ring size.

avikivity commented:
> @avikivity what do you suggest doing with this change?

Explain it better, I don't understand what it is fixing.
